examples/inspection/plot_permutation_importance_SHAP.py

```diff
@@ -0,0 +1,336 @@
+#   feature would not add up to equal to overall decrease in score from
+#   permuting all features, unless there were no interaction effects between
+#   features.
+# * Permutation importance assumes independence between features. The effect
+#   is that the 'importance' is split between correlated features. See
+#   :ref:`sphx_glr_auto_examples_inspection_plot_permutation_importance_multicollinear.py`
+#   for an example of this.
+#
+# KernelSHAP
+# ^^^^^^^^^^^
+#
+# KernelSHAP is a kernel-based method to estimate Shapley values for
+# individual samples.
+# First, it calculates the predictions for a sample when different subsets of
+# the features are 'missing'. Missingness is simulated by using a 'background'
+# (e.g., average) value for that feature. These predictions are then used to
+# fit a linear model whose predictions match that of the original model as
+# closely as possible. The coefficients of the linear 'explanation' model are
+# the Shapley values. The linear explanation model equation is:
+#
+# .. math::
+#   g(z')=\phi_0 + \sum_{i=1}^{M} \phi_i z'
+#
+# where :math:`g(z')` is the explanatory model, :math:`\phi_0` is the model
+# prediction when all features are 'missing', :math:`M` is the number of
+# possible subset sizes (n_features - 1), :math:`z'\in\{0,1\}^M`
+# (denotes the presence or absence of each feature) and :math:`\phi_i` is the
+# estimated Shapley value.
+#
+# This is much more computationally expensive than permutation importance
+# particularly as the number of possible combinations of features for all
+# feature subsets sizes quickly becomes very large as the number of features
+# increases. However, this does enable SHAP to account for interaction effects
+# between features.
+#
+# First, we will instantiate ``KernelSHAP`` using our fitted regressor,
+# ``reg`` and some background data, which will be used to simulate missingness.
+# If your dataset is small (e.g., training data is <100 samples) you can use
+# the whole training subset for the background data. Our data is larger than
+# this so we must summarize it in some way, otherwise computation will be
+# too slow. For simplicity, we will use the
+# median values of our features but SHAP offers a ``kmeans`` function that
+# can summarize each feature with ``k`` means.
+
+import shap
+
+med = X_train.median().values.reshape((1, X_train.shape[1]))
+explainer = shap.KernelExplainer(reg.predict, med)
+
+# %%
+# ``explainer`` stores various information about the data as attributes. Of
+# interest is ``expected_value``. This represents :math:`\phi_0` in our
+# equation above.
+
+explainer.expected_value
+
+# %%
+# Can you work out how this value was calculated?
+#
+# It is the prediction output by our model when using the background
+# values we gave it.
+
+reg.predict(med)
+
+# %%
+# Next we will calculate Shapley values using the testing subset. This is the
+# computationally expensive step.
+
+shap_values = explainer.shap_values(X_test, l1_reg="aic")
```

Comment:
It might be interesting to check if we can remove the warning hereby avoiding applying any regularization.

Reply:
Yes it does. The default is 'auto', which seems to use no regularization:
```
 The auto option currently uses "aic" when less that 20% of the possible sample
            space is enumerated, otherwise it uses no regularization.
```
and I can't remember why I picked 'aic' in the first place now...